tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -127,163 +128,194 @@ public void resolve() {
     typeMap = inferrer.getTypeMap();
   }
 
-  public DAGRequest buildScan(boolean idxScan) {
-    if (idxScan) {
-      return buildIndexScan();
-    } else {
-      return buildTableScan();
-    }
+  private static ColumnInfo handleColumn() {
+    return ColumnInfo.newBuilder()
+        .setColumnId(-1)
+        .setPkHandle(true)
+        // We haven't changed the field name in protobuf file, but
+        // we need to set this to true in order to retrieve the handle,
+        // so the name 'setPkHandle' may sounds strange.
+        .build();
   }
 
-  private DAGRequest buildIndexScan() {
+  /**
+   * merge buildIndexScan and buildTableScan since they are very much alike.
+   * DAGRequest for IndexScan should also contain filters and aggregation, so
+   * we can reuse this part of logic.
+   *
+   * DAGRequest is made up of a chain of executors with strict orders:
+   * TableScan/IndexScan > Selection > Aggregation > TopN/Limit
+   * DAGRequest must contain one and only one TableScan or IndexScan.
+   *
+   * @param isIndexScan whether this dagRequest is IndexScan
+   * @return final DAGRequest built
+   */
+  public DAGRequest buildScan(boolean isIndexScan) {
     checkArgument(startTs != 0, "timestamp is 0");
-    if (indexInfo == null) {
-      throw new TiClientInternalException("Index is empty for index scan");
-    }
     DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
     Executor.Builder executorBuilder = Executor.newBuilder();
     IndexScan.Builder indexScanBuilder = IndexScan.newBuilder();
+    TableScan.Builder tblScanBuilder = TableScan.newBuilder();
+    // find a column's offset in fields
+    Map<ColumnRef, Integer> colOffsetInFieldMap = new HashMap<>();
+    // find a column's position in index
+    Map<TiColumnInfo, Integer> colPosInIndexMap = new HashMap<>();
+
+    if (isIndexScan) {
+      // IndexScan
+      if (indexInfo == null) {
+        throw new TiClientInternalException("Index is empty for index scan");
+      }
+      List<TiColumnInfo> columnInfoList = tableInfo.getColumns();
+      boolean hasPk = false;
+      // We extract index column info
+      List<Integer> indexColIds = indexInfo
+          .getIndexColumns()
+          .stream()
+          .map(TiIndexColumn::getOffset)
```

Comment:
`indexColIds`, as it's name stated, we expect it contains ids of index columns, but here we use offset instead of ids, should we consider a rename here?

Reply:
like `indexColOffsets`

Reply:
done, that was the original naming